libp2p/protocols/connectivity/autorelay.nim

```diff
@@ -0,0 +1,110 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import chronos, chronicles, times, tables, sequtils, options
+import ../../switch,
+       relay/[client, utils]
+
+logScope:
+  topics = "libp2p autorelay"
+
+type
+  OnReservationHandler = proc (ma: MultiAddress): Future[void] {.gcsafe, raises: [Defect].}
+
+  AutoRelay* = ref object of Service
+    running: bool
+    runner: Future[void]
+    client: RelayClient
+    npeers: int
+    peers: Table[PeerId, Future[void]]
+    peerJoined: AsyncEvent
+    onReservation: OnReservationHandler
+
+proc reserveAndUpdate(self: AutoRelay, relayPid: PeerId, selfPid: PeerId) {.async.} =
+  while self.running:
+    let
+      rsvp = await self.client.reserve(relayPid).wait(chronos.seconds(5))
+      relayedAddr = MultiAddress.init($(rsvp.addrs[0]) &
+                                  "/p2p-circuit/p2p/" &
+                                  $selfPid).tryGet()
+    await self.onReservation(relayedAddr)
```

Comment:
can onReservation be nil?

Reply:
I thought about it but I'm not 100% sure on how to do it.
Maybe something like `method getAddresses(self: AutoRelay): seq[MultiAddress]` could be a good alternative in case there's no callback.

Reply:
I think we can check if it is `nil` before calling it, cause the service can be created like that `AutoRelay.new(3, client, nil)`

Reply:
If there is no other way to get the addresses, giving `nil` should be invalid and crash
If there is another way, we need to protect here indeed

Reply:
I added a `proc getAddresses*(self: AutoRelayService): seq[MultiAddress]` so it's no longer the only way... So it's protected now.